src/main/java/uk/gov/hmcts/reform/blobrouter/services/storage/BlobDispatcher.java

```diff
@@ -0,0 +1,53 @@
+package uk.gov.hmcts.reform.blobrouter.services.storage;
+
+import com.azure.storage.blob.BlobContainerAsyncClient;
+import com.azure.storage.blob.BlobServiceAsyncClient;
+import org.slf4j.Logger;
+import reactor.core.publisher.Flux;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.zip.ZipInputStream;
+
+import static org.slf4j.LoggerFactory.getLogger;
+
+public class BlobDispatcher {
+
+    private static final Logger logger = getLogger(BlobDispatcher.class);
+
+    private final BlobServiceAsyncClient storageClient;
+
+    public BlobDispatcher(BlobServiceAsyncClient storageClient) {
+        this.storageClient = storageClient;
+    }
+
+    public void dispatch(String blobName, ZipInputStream inputStream, String destinationContainer) throws IOException {
```

Comment:
will the caller now what to know whether this operation succeeded?

Reply:
There's `onSuccess` method which can do finalisation steps. The caller should not amend anything related to this processing route. 2 notes:

- Lease should expire automatically after 60s (max value)
- Deletion is dedicated to separate job. If not, we need to orchestrate all these calls differently by removing subscription calls and returning flux/mono to continue composition on caller side. Could also improve testing framework by allowing test framework's subscriber to watch all this

Reply:
If the upload succeeds, the caller will need to save the record in the DB. In either, success or failure, it should break the lease, as it's done working on the blob. We could theoretically pass logic to be executed on success/failure, but that would be an inelegant solution, I think.

Reply:
Any thoughts about the post-update action?

Reply:
Then suddenly this class will also need a dependency on repository layer.
I thought whatever will be calling this method will then handle the next step / call class handling the next step.
Alternatively, an action that should be called after success could be passed to this method as parameter.

Regardless, worth adding unit test.

Reply:
After posting my comment I am more leaning towards composing all actions. It'll help test and apply steps

Reply:
I'm not that much agains passing logic - we can see. But I think we should pass a byte array into this method so that:
- this method doesn't synchronously load the blob before asynchronously uploading it
- we don't have any zip-specific failure in this method - at this point we know we want to upload it and verification should happen earlier